ably/realtime_presence.go

```diff
@@ -0,0 +1,230 @@
+package ably
+
+import (
+	"errors"
+	"sync"
+
+	"github.com/ably/ably-go/ably/proto"
+)
+
+// RealtimePresence represents a single presence map of a prarticular channel.
+// It allows entering, leaving and updating presence state for the current
+// client or on behalf of other client.
+type RealtimePresence struct {
+	mu      sync.Mutex
+	sync    sync.WaitGroup
+	data    string
+	serial  string
+	subs    *subscriptions
+	channel *RealtimeChannel
+	members map[string]*proto.PresenceMessage
+	stale   map[string]struct{}
+	state   proto.PresenceState
+	insync  bool
+	synced  bool
+}
+
+func newRealtimePresence(channel *RealtimeChannel) *RealtimePresence {
+	return &RealtimePresence{
+		subs:    newSubscriptions(subscriptionPresenceMessages),
+		channel: channel,
+		members: make(map[string]*proto.PresenceMessage),
+	}
+}
+
+func (pres *RealtimePresence) send(msg *proto.PresenceMessage, result bool) (Result, error) {
+	if _, err := pres.channel.attach(false); err != nil {
+		return nil, err
+	}
+	switch state := pres.channel.State(); state {
+	case StateChanDetached, StateChanDetaching, StateChanClosing, StateChanClosed,
+		StateChanFailed:
+		return nil, newError(91001, errors.New(StateText(state)))
+	}
+	protomsg := &proto.ProtocolMessage{
+		Action:   proto.ActionPresence,
+		Channel:  pres.channel.state.channel,
+		Presence: []*proto.PresenceMessage{msg},
+	}
+	return pres.channel.send(protomsg, result)
+}
+
+func (pres *RealtimePresence) syncStart() {
+	if pres.insync {
+		return
+	}
+	pres.sync.Add(1)
+	pres.insync = true
```

Comment:
`insync` is a bit misleading, I assume `insync` means a sync is in progress and not that the Presence Map is in sync.  Therefore, could we perhaps name it something like `syncInProgress` or `syncComplete` and set it to false?